Add chains to build component plus related scripts

simonbaird opened the pull request:
Includes:
- New gitops config to install and configure tekton chains
- Bash scripts for:
    - Creating a key-pair for chains signing
    - Doing some CA cert juggling to make SSL work
    - Demoing chains with cosign and rekor verification

JIRA: [HACBS-38](https://issues.redhat.com/browse/HACBS-38)
JIRA: [HACBS-47](https://issues.redhat.com/browse/HACBS-47)

simonbaird commented:
It's working for me in my local CRC cluster.

simonbaird commented:
An earlier version of this PR is at https://github.com/simonbaird/infra-deployments/pull/1 .

ralphbean commented:
/ok-to-test

simonbaird commented on hack/bootstrap-cluster.sh:
```diff
@@ -51,6 +51,11 @@ kubectl apply -f $ROOT/argo-cd-apps/app-of-apps/all-applications-staging.yaml
+echo
+echo "Configuring service account for Tekton Chains:"
+oc new-project tekton-chains
+oc adm policy add-scc-to-user anyuid -z tekton-chains-controller
```
I did wonder about that, but I couldn't figure out how to do it.

Michkov replied:
The problem is here https://github.com/tektoncd/chains/blob/main/config/100-deployment.yaml#L103-L105, they are trying to secure kubernetes and that breaks openshift...

I am able to workaround this by kustomize patchesStrategicMerge:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: tekton-chains-controller
  namespace: tekton-chains
spec:
  template:
    spec:
      containers:
      - name: tekton-chains-controller
        securityContext:
          runAsGroup:
          runAsUser:
```

simonbaird replied:
Thanks, I'll give it a try.

simonbaird commented on hack/chains/kaniko-demo-build.sh:
```diff
@@ -0,0 +1,32 @@
+kubectl patch task/kaniko-chains --type=json \
+  --patch='[{"op": "add", "path": "/spec/steps/1/args/5", "value":"--skip-tls-verify=true"}]'
+
+REGISTRY=default-route-openshift-image-registry.apps-crc.testing
```
Good idea, thanks. For quay.io, I think that's out of scope for this PR, should be something we can look at soon.

simonbaird replied:
Thinking about this some more -- For the cosign part of the demo, currently cosign is run locally and needs access to the registry. Is there a way to expose the registry using cluster bot? If not then maybe the cosign demo as it is now can't work.

Michkov replied:
It's possible (https://docs.openshift.com/container-platform/4.9/registry/securing-exposing-registry.html) but the question is if we want to do it. The environment should be very similar to staging/prod where I suppose the registry will not be exposed. Also the scripts should be demo-able on stage instance. Integration with quay.io would solve this and would remove that cert configuration.

simonbaird commented:
I have a few changes coming, so I'll move this back to draft.

simonbaird commented:
I wasn't able to get the gitops change working in a cluster-bot cluster, but I think maybe it's unrelated to this patch. After running hack/preview.sh the Argo CD would not sync and I could see an error like this:

```
ComparisonError: rpc error: code = Internal desc = Failed to checkout ee855d9155bd6d1830939577b7e4d3e9b1edce3c: `git checkout ...
```

I try to figure out my cluster bot problem later, but for now I think this can be reviewed again.

The PR addresses suggestions and review comments, so I'll move it out of draft. Notable changes:

- Remove the `oc adm policy add-scc-to-user anyuid ...` command. I've done that in its own commit, see "Modify chains security..."
- Tweak the traskrun or pipelinerun name handling in some bash scripts so make it more flexible
- Output some more rekor data in the rekor demo

Michkov commented on hack/bootstrap-cluster.sh:
```diff
@@ -67,3 +67,5 @@ echo "OK"
 echo "Login/password uses your OpenShift credentials ('Login with OpenShift' button)"
 echo
+
+oc project default
```
not needed.

simonbaird replied:
You're right, I'll remove it. The motivation was to avoid landing on tekton-chains, which isn't happening any more.

simonbaird replied:
It's removed in the latest revision.

Michkov commented on hack/chains/setup-controller-certs.sh:
```diff
@@ -0,0 +1,70 @@
+#
+
+# A cert from your local cluster to get local SSL working
+kubectl get secret router-ca -n openshift-ingress-operator -o jsonpath="{.data.tls\.crt}" | base64 -d > $TMP_DIR/cluster-ingress.pem
```
Does not work with cluster bot, internal network is using different certificate. 
```
getting signed image: Get \"https://image-registry.openshift-image-registry.svc:5000/v2/\": x509: certificate signed by unknown authority
```

simonbaird replied:
Did the gitops change work for you in cluster bot? Perhaps I should split that into a separate PR so it could be merged sooner.

simonbaird replied:
Ack re the CA cert. Do you know where to find the right cert?

Michkov commented:
@simonbaird I also see that sync issue but only with your changes, when I do bootstrap and preview in main branch then I do not see that sync issue.

simonbaird commented:
> @simonbaird I also see that sync issue but only with your changes, when I do bootstrap and preview in main branch then I do not see that sync issue.

Any idea what might be causing it?

simonbaird commented on hack/chains/check-pipelinerun.sh:
```diff
@@ -2,22 +2,14 @@
 TASKRUN_NAMES=$(
-  kubectl get pipelinerun/$PIPELINERUN_NAME -o yaml |
-    yq e '.status.taskRuns | keys | .[]' - )
+  kubectl get $PIPELINERUN_NAME -o yaml | yq e '.status.taskRuns | keys | .[]' - )
```
Fixed in the newer revisions.

simonbaird commented:
I'll try again with the cluster-bot deploy.

Michkov commented:
@simonbaird please check #109 , when I rebased your change to that MR I was able to deploy without sync issue. It's simplifying deployment via argocd since it's going directly to preview branch. But still not sure if that sync issue is random or not.

simonbaird commented:
Let's move back to draft.

simonbaird commented:
Thanks @Michkov, I'll try it.

sbose78 commented on components/build/tekton-chains/kustomization.yaml:
```diff
@@ -0,0 +1,13 @@
+resources:
+# Later we might want to pin this to a version, or download it and split it into
+# its separate pieces, but for now let's just use the official release directly
+- https://storage.googleapis.com/tekton-releases/chains/latest/release.yaml
```
I've been burnt with sudden changes in need for privileges - wondering if this a better option ?  https://storage.googleapis.com/tekton-releases/chains/previous/v0.8.0/release.yaml

sbose78 replied:
does this have a specific fix we are carrying?

simonbaird replied:
I'm happy to set it to 0.8.0 for stability if that's preferred.

Michkov replied:
I am for at least selection of the version in url, so the versioning will be under our control.

sbose78 commented on components/build/tekton-chains/security-context-fix.yaml:
```diff
@@ -0,0 +1,27 @@
+      - name: tekton-chains-controller
+        securityContext:
+          runAsGroup:
+          runAsUser:
```
Didn't we require runAsUser: 0 because of a latest issue, or has that been fixed upstream?

simonbaird replied:
I'm not aware of that, but I'll do some googling.

sbose78 commented on hack/chains/setup-controller-certs.sh:
```diff
@@ -0,0 +1,70 @@
+#kubectl get secret chains-ca-cert -n tekton-chains -o jsonpath='{.data.ca-certificates\.crt}' | base64 -d
+
+#
+# Now make the certs accessible where they're needed
```
Does any of this help ? https://docs.openshift.com/container-platform/4.6/security/certificates/service-serving-certificate.html

simonbaird replied:
Thanks, I'll do some reading.

sbose78 reviewed (commented):
As a follow-up after merging this, we'd need to create the signing secret too in the `tekton-chains` namespace too, right ?

sbose78 commented:
Apart from creation of the signing secret, is there anything that needs to be done with manual cluster access ?

simonbaird commented:
> As a follow-up after merging this, we'd need to create the signing secret too in the `tekton-chains` namespace too, right ?

That's right. You could run hack/chains/create-signing-secret.sh to do it, but I was wondering about the idea of having it added automatically by Argo CD somehow.

simonbaird commented:
Notable changes in this revision:
* Pin the chains version to 0.0.8
* Rebase on the latest upstream/main

simonbaird commented:
FYI this is likely to be abandoned in favor of https://github.com/hacbs-contract/infra-deployments/pull/1 .

The idea is we (contract team) will be able to collaborate better if we merge things into a working branch that we can pull down and build on top of. When we think it's ready to merge back into main there'll be another PR to review.

simonbaird commented:
https://github.com/hacbs-contract/infra-deployments/pull/1 is merged now to the new 'contract-devel' branch FYI.

In the future there'll me a new PR to merge that branch to main which will include a version of this change. Abandoning this PR now.

simonbaird commented:
Thanks for the reviews and suggestions. Improvements based on the CA certs, using external registries like quay.io, signing secret management, improved demos, clusterbot fixes, etc should be visible as PRs over at https://github.com/hacbs-contract/infra-deployments/pulls .
